# Relative quantification proof of concept

Take a first pass at implementing evaluation of (fractional) relative quantification, more or less following [Joseph's blueprint](https://docs.google.com/document/d/1u18BSWB2776d-EXAoft7Je5KmhC1aGMFZJRpJ1rueNI/edit) with a [few practical pointers I had come up with](https://docs.google.com/document/d/1t30SEHZ6RSFYXG7o1ylWCfgYWVbZCA1jj-SkES1M2rw/edit)

Purpose isn't to necessarily to make something fully flexible/ready to go, but to explore how to implement the blueprint and understand any impacts our filtering criteria/decisions may have.


Broadly, this involves:
- Defining list of genes expressed above a given TPM threshold from RNA-seq data
- Defining a set of terminal exons w/ 2 polyA sites from the GT poly(A) site BEDs, filtering to 2 most expressed sites if a TE has > 2 PAS
- Filtering the predicted sites to those occurring on curated set of terminal exons
- 


Will focus on a single Mayr sample Mayr_CD5B_R3 with outputs from DaPars2 & DaPars for now...

In [4]:
import pyranges as pr
import pandas as pd
import numpy as np
import os
import warnings

# There are some FutureWarnings with installed version of PyRanges and Pandas - For visual purposes just don't print them for now
warnings.filterwarnings('ignore')

# 1. Find genes expressed in RNA-seq above a given threshold


Matt provided a TSV of Salmon quantification outputs from the nf-core RNA-seq workflow (**TODO: insert link**). As this is per-transcript, need to sum to gene-level

Target is a list/set of Ensembl gene IDs that pass a given TPM threshold

In [5]:
tpm_df = pd.read_csv("data/Mayr_salmon.merged.transcript_tpm.tsv", sep="\t")
tpm_df.head(n=10)

tx            gene_id  Mayr_CD5B_R3  Mayr_CD5B_R4  \
0  ENST00000456328.2  ENSG00000223972.5       0.00000      0.044865   
1  ENST00000450305.2  ENSG00000223972.5       0.00000      0.000000   
2  ENST00000488147.1  ENSG00000227232.5       2.62799      3.314172   
3  ENST00000619216.1  ENSG00000278267.1       0.00000      0.000000   
4  ENST00000473358.1  ENSG00000243485.5       0.00000      0.000000   
5  ENST00000469289.1  ENSG00000243485.5       0.00000      0.000000   
6  ENST00000607096.1  ENSG00000284332.1       0.00000      0.000000   
7  ENST00000417324.1  ENSG00000237613.2       0.00000      0.000000   
8  ENST00000461467.1  ENSG00000237613.2       0.00000      0.000000   
9  ENST00000606857.1  ENSG00000268020.3       0.00000      0.000000   

   Mayr_GC_R1  Mayr_GC_R2  Mayr_GC_R3  Mayr_GC_R4  Mayr_M_R2  Mayr_M_R6  \
0    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
1    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
2    2.877079    2.625817    2.839663    3.452287   2.285098   2.630034   
3    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
4    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
5    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
6    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
7    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
8    0.000000    0.026530    0.000000    0.000000   0.030134   0.000000   
9    0.000000    0.000000    0.000000    0.000000   0.073951   0.000000   

   Mayr_NB_R1  Mayr_NB_R2  Mayr_NB_R3  Mayr_NB_R4  Mayr_NB_R5  Mayr_NB_R6  
0    0.135770    0.012529    0.000000    0.000000    0.000000     0.00000  
1    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000  
2    2.627121    2.015704    2.529293    1.699014    2.747853     2.22484  
3    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000  
4    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000  
5    0.000000    0.000000    0.000000    0.000000    0.000000     0.03989  
6    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000  
7    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000  
8    0.000000    0.000000    0.026825    0.000000    0.000000     0.00000  
9    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000

In [6]:
# Summarise individual transcript expression to gene-level TPM
sample_cols = [col for col in tpm_df.columns if col not in ["tx", "gene_id"]]
# sample_cols

gene_tpm = tpm_df.groupby("gene_id")[sample_cols].sum()
gene_tpm

Mayr_CD5B_R3  Mayr_CD5B_R4  Mayr_GC_R1  Mayr_GC_R2  \
gene_id                                                                  
ENSG00000000003.15      0.841496      0.313483    0.224971    0.220765   
ENSG00000000005.6       0.000000      0.000000    0.000000    0.000000   
ENSG00000000419.14     31.462058     27.354208   35.251502   29.225256   
ENSG00000000457.14     13.687977     13.619168    6.316804    5.554757   
ENSG00000000460.17      6.285515      4.806302   22.109690   18.177691   
...                          ...           ...         ...         ...   
ENSG00000288721.1       0.299191      0.590024    0.962898    1.188937   
ENSG00000288722.1       3.003168      0.794827    4.912289    2.726699   
ENSG00000288723.1       0.032072      0.000000    0.000000    0.000000   
ENSG00000288724.1       0.000000      0.000000    0.000000    0.000000   
ENSG00000288725.1       1.517664      1.601146    1.266355    1.549774   

                    Mayr_GC_R3  Mayr_GC_R4  Mayr_M_R2  Mayr_M_R6  Mayr_NB_R1  \
gene_id                                                                        
ENSG00000000003.15    0.022432    0.569883   0.789282   0.378343    0.021219   
ENSG00000000005.6     0.000000    0.000000   0.000000   0.000000    0.000000   
ENSG00000000419.14   24.685265   34.762234  28.490978  18.881510   35.288736   
ENSG00000000457.14    4.834584    5.143130   9.068360   5.792793    8.083517   
ENSG00000000460.17   14.991760   14.880926   4.207996   4.209314    1.514664   
...                        ...         ...        ...        ...         ...   
ENSG00000288721.1     0.839328    1.087980   0.350958   0.304280    0.505675   
ENSG00000288722.1     1.875441    7.280854   0.602439   2.423762    1.573895   
ENSG00000288723.1     0.000000    0.000000   0.029104   0.000000    0.000000   
ENSG00000288724.1     0.000000    0.000000   0.000000   0.000000    0.000000   
ENSG00000288725.1     1.635822    2.798908   2.485099   1.508843    0.841387   

                    Mayr_NB_R2  Mayr_NB_R3  Mayr_NB_R4  Mayr_NB_R5  Mayr_NB_R6  
gene_id                                                                         
ENSG00000000003.15    0.000000    1.269005    0.835454    0.435978    0.007870  
ENSG00000000005.6     0.000000    0.000000    0.053663    0.000000    0.000000  
ENSG00000000419.14   26.330158   24.401633   22.140657   19.653655   20.561543  
ENSG00000000457.14    0.864191    8.795230    7.967781    6.033806    1.219366  
ENSG00000000460.17    0.489746    5.427459    3.280607    3.677558    0.469440  
...                        ...         ...         ...         ...         ...  
ENSG00000288721.1     0.029361    0.246273    0.191482    0.293596    0.159076  
ENSG00000288722.1     1.523601    4.048628    0.883793    2.006525    1.103560  
ENSG00000288723.1     0.000000    0.000000    0.000000    0.000000    0.000000  
ENSG00000288724.1     0.000000    0.000000    0.000000    0.000000    0.000000  
ENSG00000288725.1     0.000000    0.612817    0.855837    0.000000    0.000000  

[60649 rows x 14 columns]

In [7]:
# Create a matrix of Booleans for each column for whether gene passes expression filter
thresh_tpm1 = gene_tpm > 1
thresh_tpm1

Mayr_CD5B_R3  Mayr_CD5B_R4  Mayr_GC_R1  Mayr_GC_R2  \
gene_id                                                                  
ENSG00000000003.15         False         False       False       False   
ENSG00000000005.6          False         False       False       False   
ENSG00000000419.14          True          True        True        True   
ENSG00000000457.14          True          True        True        True   
ENSG00000000460.17          True          True        True        True   
...                          ...           ...         ...         ...   
ENSG00000288721.1          False         False       False        True   
ENSG00000288722.1           True         False        True        True   
ENSG00000288723.1          False         False       False       False   
ENSG00000288724.1          False         False       False       False   
ENSG00000288725.1           True          True        True        True   

                    Mayr_GC_R3  Mayr_GC_R4  Mayr_M_R2  Mayr_M_R6  Mayr_NB_R1  \
gene_id                                                                        
ENSG00000000003.15       False       False      False      False       False   
ENSG00000000005.6        False       False      False      False       False   
ENSG00000000419.14        True        True       True       True        True   
ENSG00000000457.14        True        True       True       True        True   
ENSG00000000460.17        True        True       True       True        True   
...                        ...         ...        ...        ...         ...   
ENSG00000288721.1        False        True      False      False       False   
ENSG00000288722.1         True        True      False       True        True   
ENSG00000288723.1        False       False      False      False       False   
ENSG00000288724.1        False       False      False      False       False   
ENSG00000288725.1         True        True       True       True       False   

                    Mayr_NB_R2  Mayr_NB_R3  Mayr_NB_R4  Mayr_NB_R5  Mayr_NB_R6  
gene_id                                                                         
ENSG00000000003.15       False        True       False       False       False  
ENSG00000000005.6        False       False       False       False       False  
ENSG00000000419.14        True        True        True        True        True  
ENSG00000000457.14       False        True        True        True        True  
ENSG00000000460.17       False        True        True        True       False  
...                        ...         ...         ...         ...         ...  
ENSG00000288721.1        False       False       False       False       False  
ENSG00000288722.1         True        True       False        True        True  
ENSG00000288723.1        False       False       False       False       False  
ENSG00000288724.1        False       False       False       False       False  
ENSG00000288725.1        False       False       False       False       False  

[60649 rows x 14 columns]

In [8]:
# Number of genes passing threshold in each sample
thresh_tpm1.sum(axis=0)

Mayr_CD5B_R3    12718
Mayr_CD5B_R4    12288
Mayr_GC_R1      12662
Mayr_GC_R2      12436
Mayr_GC_R3      11942
Mayr_GC_R4      12673
Mayr_M_R2       12103
Mayr_M_R6       12321
Mayr_NB_R1      11639
Mayr_NB_R2       9477
Mayr_NB_R3      12245
Mayr_NB_R4      11996
Mayr_NB_R5      12070
Mayr_NB_R6       9645
dtype: int64

In [9]:
# Fraction of annotated genes passing threshold in each sample
# NB: could subset to prot-coding & lncRNA genes (as expect their RNAs to be polyadenylated)
# don't think this is that informative as don't expect all annotated genes to be expressed...
thresh_tpm1.sum(axis=0) / len(thresh_tpm1)

Mayr_CD5B_R3    0.209698
Mayr_CD5B_R4    0.202608
Mayr_GC_R1      0.208775
Mayr_GC_R2      0.205049
Mayr_GC_R3      0.196903
Mayr_GC_R4      0.208956
Mayr_M_R2       0.199558
Mayr_M_R6       0.203153
Mayr_NB_R1      0.191908
Mayr_NB_R2      0.156260
Mayr_NB_R3      0.201899
Mayr_NB_R4      0.197794
Mayr_NB_R5      0.199014
Mayr_NB_R6      0.159030
dtype: float64

In [10]:
# Dict of {sample_name: {gene_ids}} that pass the expression filter
genes_thresh_tpm1 = {col: set(thresh_tpm1.index[thresh_tpm1.loc[:, col]]) for col in thresh_tpm1.columns}

# 2 - Find / Decide TEs to select for filtering ground truth

- Merge/'union' overlapping TEs to create the longest possible TE isoform for searching in GT for overlapping sites (most permissive)
    - In cases where TE isoforms have different 3'ss, they will be merged regardless (i.e. not considered as separate events)



In [11]:
%%time
gtf = pr.read_gtf("data/gencode.v38.annotation.gtf")


CPU times: user 1min 14s, sys: 4.43 s, total: 1min 18s
Wall time: 1min 19s


In [12]:
# PyRanges objects are dictionaries of pandas dataframes
# Keys are chrom names & strand, values are dfs storing genomic coords and metadata
gtf.head()

+--------------+------------+------------+-----------+-----------+-------+
| Chromosome   | Source     | Feature    |     Start |       End | +20   |
| (category)   | (object)   | (object)   |   (int32) |   (int32) | ...   |
|--------------+------------+------------+-----------+-----------+-------|
| chr1         | HAVANA     | gene       |     11868 |     14409 | ...   |
| chr1         | HAVANA     | transcript |     11868 |     14409 | ...   |
| chr1         | HAVANA     | exon       |     11868 |     12227 | ...   |
| chr1         | HAVANA     | exon       |     12612 |     12721 | ...   |
| chr1         | HAVANA     | exon       |     13220 |     14409 | ...   |
| chr1         | HAVANA     | transcript |     12009 |     13670 | ...   |
| chr1         | HAVANA     | exon       |     12009 |     12057 | ...   |
| chr1         | HAVANA     | exon       |     12178 |     12227 | ...   |
+--------------+------------+------------+-----------+-----------+-------+
Stranded PyRanges object has 8 rows and 25 columns from 1 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
20 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 13 more.)

In [13]:
gtf.stranded

True

In [35]:
gtf = gtf[["Feature", "gene_id", "transcript_id", "gene_name", "exon_number", "gene_type", "transcript_type", "transcript_support_level"]]

gtf

+--------------+------------+-----------+-----------+--------------+-------+
| Chromosome   | Feature    | Start     | End       | Strand       | +7    |
| (category)   | (object)   | (int32)   | (int32)   | (category)   | ...   |
|--------------+------------+-----------+-----------+--------------+-------|
| chr1         | gene       | 11868     | 14409     | +            | ...   |
| chr1         | transcript | 11868     | 14409     | +            | ...   |
| chr1         | exon       | 11868     | 12227     | +            | ...   |
| chr1         | exon       | 12612     | 12721     | +            | ...   |
| ...          | ...        | ...       | ...       | ...          | ...   |
| chrY         | exon       | 57213879  | 57213964  | -            | ...   |
| chrY         | exon       | 57213525  | 57213602  | -            | ...   |
| chrY         | exon       | 57213203  | 57213357  | -            | ...   |
| chrY         | exon       | 57212183  | 57213125  | -            | ...   |
+--------------+------------+-----------+-----------+--------------+-------+
Stranded PyRanges object has 3,150,424 rows and 12 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
7 hidden columns: gene_id, transcript_id, gene_name, exon_number, gene_type, transcript_type, ... (+ 1 more.)

In [36]:
# For simplicities sake let's focus on a single sample
# Mayr_CD5B_R3
exons = gtf.subset(lambda df: df["Feature"] == "exon")
# Need as int dtype for some downstream functions to work
exons = exons.assign("exon_number", lambda df: df["exon_number"].astype(float).astype(int))

exons.dtypes


Chromosome                  category
Feature                       object
Start                          int32
End                            int32
Strand                      category
gene_id                       object
transcript_id                 object
gene_name                     object
exon_number                    int64
gene_type                     object
transcript_type               object
transcript_support_level      object
dtype: object

In [37]:
exons_mayr_cd_r3_tpm_filt = exons.subset(lambda df: df["gene_id"].isin(genes_thresh_tpm1["Mayr_CD5B_R3"]))

exons_mayr_cd_r3_tpm_filt

+--------------+------------+-----------+-----------+--------------+-------+
| Chromosome   | Feature    | Start     | End       | Strand       | +7    |
| (category)   | (object)   | (int32)   | (int32)   | (category)   | ...   |
|--------------+------------+-----------+-----------+--------------+-------|
| chr1         | exon       | 629061    | 629433    | +            | ...   |
| chr1         | exon       | 629639    | 630683    | +            | ...   |
| chr1         | exon       | 631073    | 632616    | +            | ...   |
| chr1         | exon       | 633695    | 634376    | +            | ...   |
| ...          | ...        | ...       | ...       | ...          | ...   |
| chrY         | exon       | 20570221  | 20570519  | -            | ...   |
| chrY         | exon       | 21044598  | 21044724  | -            | ...   |
| chrY         | exon       | 21042268  | 21042369  | -            | ...   |
| chrY         | exon       | 21038288  | 21039044  | -            | ...   |
+--------------+------------+-----------+-----------+--------------+-------+
Stranded PyRanges object has 894,995 rows and 12 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
7 hidden columns: gene_id, transcript_id, gene_name, exon_number, gene_type, transcript_type, ... (+ 1 more.)

In [16]:
# Extract terminal exons for each transcript
def get_terminal_exons(gr,
                         feature_col = "Feature",
                         feature_key = "exon",
                         id_col = "transcript_id",
                         region_number_col = "exon_number",
                         filter_single = False
                         ):
    '''
    Return gr of last exons for each transcript
    
    Function makes the following assumptions:
    - region_number_col values define non-overlapping regions (exons) of the same group (transcript_id)
    - region_number_col is strand aware 5'-3' in ascending order
        - i.e. 1st exon (most 5') is numbered 1, last exon of transcript with n exons is labelled n
        - For reference GTFs this is probably a fair assumption
    - ONLY a single unique value (feature_key) is present in feature_col (e.g. only 'exon' present)
    - region_number_col is an int np.dtype col
    - regions are not duplicated within the gr
    
    Notes:
    - use add_region_number if not confident region_number_col matches expectation
        
    '''

    assert region_number_col in gr.columns.tolist()
    assert feature_col in gr.columns.tolist()
    assert id_col in gr.columns.tolist()

    # Make sure region_number is an int dtype (so can be sorted numerically)
    assert gr.dtypes.loc[region_number_col] in [np.dtype('int32'), np.dtype('int64')]
    # Make sure only 'exon' features are in the gr
    assert gr.as_df()[feature_col].drop_duplicates().tolist() == [feature_key], "only {} entries should be present in gr".format(feature_key)


    # Make sure gr is sorted by transcript_id & 'region number' (ascending order so 1..n)
    mod_gr = gr.apply(lambda df: df.sort_values(by=[id_col, region_number_col], ascending=True),
                          nb_cpu=1)


    # Filter out single-exon transcripts
    if filter_single:
        print("Filtering for multi-exon transcripts...")
        print(f"Before: {len(set(mod_gr.as_df()[id_col]))}")

        # Setting to 'False' marks all duplicates as True (so keeps transcript IDs with multiple exons these)
        mod_gr = mod_gr.subset(lambda df: df.duplicated(subset=[id_col], keep=False), nb_cpu=1)

        eprint(f"After: {len(set(mod_gr.as_df()[id_col]))}")

    # Pick last region entry by max region number for each transcript (id_col)
    # keep="last" sets last in ID to 'False' and all others true (negate to keep last only)

    out_gr = mod_gr.subset(lambda df: ~(df.duplicated(subset=[id_col], keep="last")),
                           nb_cpu=1
                           )

    # re-sort by genomic coords not transcript_id
    return out_gr.sort()


def _df_add_region_number(df, id_col, sort_col="Start"):
    '''
    Return a Series of strand-aware region numbers (5'-3' in 1..n)
    Function to be used internally in a pr.assign (mainly by add_region_number)
    '''
    if id_col not in df.columns.tolist():
        raise KeyError(f"id_col - {id_col} - is not present in df for chr/strand pair {','.join([df.Chromosome.iloc[0], df.Strand.iloc[0]])}")

    elif (df.Strand == "+").all():
        # Start position smallest to largest = 5'-3'

        return df.groupby(id_col)[sort_col].rank(method="min", ascending=True)

    elif (df.Strand == "-").all():
        # Start position largest to smallest = 5'-3'

        return df.groupby(id_col)[sort_col].rank(method="min", ascending=False)

    elif df.empty:
        eprint("df is empty - returning empty pd.Series")
        return pd.Series()


def add_region_number(gr,
                      id_col="transcript_id",
                      out_col="exon_number",
                      feature_col="Feature",
                      feature_key="exon",
                      ):
    '''
    Adds column to gr containing a strand aware region number column,
    ordered 5'-3' 1..n by a group of features (e.g. transcript)
    
    Assumes every region/row of feature in group is not duplicated (i.e. given exon only appears once)
    '''

    # Make sure only 'feature_key' rows are in the gr
    assert gr.as_df()[feature_col].drop_duplicates().tolist() == [feature_key], "only {} entries should be present in gr".format(feature_key)

    # Make sure sorted by position first.
    gr = gr.sort()

    # Add in region number column in strand aware manner, so 1 = most 5', n = most 3'

    gr = gr.assign(out_col, lambda df: _df_add_region_number(df, id_col))

    return gr

In [38]:
te_mayr_cd_r3_tpm_filt = get_terminal_exons(exons_mayr_cd_r3_tpm_filt)

te_mayr_cd_r3_tpm_filt

+--------------+------------+-----------+-----------+--------------+-------+
| Chromosome   | Feature    | Start     | End       | Strand       | +7    |
| (category)   | (object)   | (int32)   | (int32)   | (category)   | ...   |
|--------------+------------+-----------+-----------+--------------+-------|
| chr1         | exon       | 629061    | 629433    | +            | ...   |
| chr1         | exon       | 629639    | 630683    | +            | ...   |
| chr1         | exon       | 631073    | 632616    | +            | ...   |
| chr1         | exon       | 633695    | 634376    | +            | ...   |
| ...          | ...        | ...       | ...       | ...          | ...   |
| chrY         | exon       | 20504475  | 20507456  | -            | ...   |
| chrY         | exon       | 20506768  | 20507456  | -            | ...   |
| chrY         | exon       | 20570221  | 20570519  | -            | ...   |
| chrY         | exon       | 21038288  | 21039044  | -            | ...   |
+--------------+------------+-----------+-----------+--------------+-------+
Stranded PyRanges object has 119,902 rows and 12 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
7 hidden columns: gene_id, transcript_id, gene_name, exon_number, gene_type, transcript_type, ... (+ 1 more.)

In [39]:
# Merge TEs into non-overlapping, union TEs
m_te_mayr_cd_r3_tpm_filt = te_mayr_cd_r3_tpm_filt.merge(strand=True, by="gene_id")
m_te_mayr_cd_r3_tpm_filt

+--------------+-----------+-----------+--------------+--------------------+
| Chromosome   | Start     | End       | Strand       | gene_id            |
| (category)   | (int32)   | (int32)   | (category)   | (object)           |
|--------------+-----------+-----------+--------------+--------------------|
| chr1         | 169804074 | 169804386 | +            | ENSG00000000460.17 |
| chr1         | 169807790 | 169807837 | +            | ENSG00000000460.17 |
| chr1         | 169821678 | 169821719 | +            | ENSG00000000460.17 |
| chr1         | 169852789 | 169854080 | +            | ENSG00000000460.17 |
| ...          | ...       | ...       | ...          | ...                |
| chrY         | 11153857  | 11154070  | -            | ENSG00000278212.2  |
| chrY         | 11155653  | 11155884  | -            | ENSG00000278212.2  |
| chrY         | 11162449  | 11163137  | -            | ENSG00000278212.2  |
| chrY         | 12661308  | 12662124  | -            | ENSG00000286009.2  |
+--------------+-----------+-----------+--------------+--------------------+
Stranded PyRanges object has 48,902 rows and 5 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.

In [42]:
%%time
# If want to retain metadata (e.g. transcript IDs) for overlapping TEs, can use functions below
# This approach is a fair bit slower so won't use ds (as just need merged regions), but relevant for Matt's suggestion of a terminal exon ID
# could probably speed up with nb_cpu for multiple processors...
def _collapse_tes(df,
                  cluster_col,
                  distinct_cols
                  ):
    '''
    '''
 
    col_order = df.columns.tolist()
            
    other_cols = [col for col in col_order if col not in distinct_cols]
    # First define how want to collapse metadata columns to single row values
    # 
    # Cols want to combine values as ; separated string (e.g. if expect diff values per row)
    cat_dict = {col: lambda x: x.astype(str).str.cat(sep=";") 
                  for col in df.columns
                    if col not in distinct_cols}

    # Cols want to report a single value (e.g. where expect same values in all rows/not important)
    # Also important where need to maintain original 
    distinct_dict = {col: lambda x: x.first()
                         for col in distinct_cols
                        }
    
    # Define the merged ranges
    cat_dict["Start"] = lambda x: x.min()
    cat_dict["End"] = lambda x: x.max()
    
    grpd = df.groupby(cluster_col, group_keys=False)
    
    grpd_distinct = grpd[distinct_cols].first()
    grpd_other = grpd[other_cols].agg(cat_dict)
    
    # Combine to single row
    clpsd = grpd_distinct.merge(grpd_other, left_index=True, right_index=True)
    
    return clpsd[col_order]


def collapse_tes(gr,
                 cluster_col="Cluster",
                 distinct_cols=["Chromosome", "Feature", "gene_id", "Strand", "gene_name"]):
    
    # Want to keep grouping col to single value
    dist_cols = distinct_cols + [cluster_col]
    
    return gr.apply(lambda df: _collapse_tes(df, cluster_col, dist_cols))

# i = 0 
# for _, df in tmp_grp:
#     while i < 2:
#         print(df)
#         tmp_df = df.copy()
#         cat_dict = {col: lambda x: x.astype(str).str.cat(sep=";") 
#                     for col in tmp_df.columns
#                     if col not in ["Chromosome", "Start", "End", "Feature", "gene_id", "Strand"]}

#         distinct_dict = {col: lambda x: x.head(n=1)
#                          for col in ["Chromosome", "Feature", "gene_id", "Strand"]
#                         }
        
#         agg_dict = {** cat_dict, ** distinct_dict}
        
#         agg_dict["Start"] = lambda x: x.min()
#         agg_dict["End"] = lambda x: x.max()
        
#         print(df.agg(agg_dict))
                
#         i +=1
#     else:
#         break
        
collapse_tes(te_mayr_cd_r3_tpm_filt.cluster(strand=True, by="gene_id"))

CPU times: user 53.7 s, sys: 11.7 ms, total: 53.7 s
Wall time: 53.7 s


+--------------+------------+-----------+-----------+--------------+-------+
| Chromosome   | Feature    | Start     | End       | Strand       | +8    |
| (category)   | (object)   | (int32)   | (int32)   | (category)   | ...   |
|--------------+------------+-----------+-----------+--------------+-------|
| chr1         | exon       | 169804074 | 169804386 | +            | ...   |
| chr1         | exon       | 169807790 | 169807837 | +            | ...   |
| chr1         | exon       | 169821678 | 169821719 | +            | ...   |
| chr1         | exon       | 169852789 | 169854080 | +            | ...   |
| ...          | ...        | ...       | ...       | ...          | ...   |
| chrY         | exon       | 11153857  | 11154070  | -            | ...   |
| chrY         | exon       | 11155653  | 11155884  | -            | ...   |
| chrY         | exon       | 11162449  | 11163137  | -            | ...   |
| chrY         | exon       | 12661308  | 12662124  | -            | ...   |
+--------------+------------+-----------+-----------+--------------+-------+
Stranded PyRanges object has 48,902 rows and 13 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
8 hidden columns: gene_id, transcript_id, gene_name, exon_number, gene_type, transcript_type, ... (+ 2 more.)

In [44]:
# To do/check: if we have unstranded RNA_seq data, consider removing TEs that have any overlap with TEs of diff genes?

(m_te_mayr_cd_r3_tpm_filt.join(m_te_mayr_cd_r3_tpm_filt,
                               how="left",
                               )
 .subset(lambda df: df["gene_id"] != df["gene_id_b"])
 .as_df()
 [["Strand", "Strand_b"]].value_counts()
)

Strand  Strand_b
+       -           799
-       +           799
+       +           332
-       -           292
dtype: int64

**NB: I don't think above is massively important but keeping for completeness**

A very small minority of TEs overlap with TEs of other gene_ids... Most of these overlaps occur on opposite strands, so algos that can profile txpn strand & prediction datasets that are strand-specific should be able to differentiate.

TODO: check overlap of TEs with other exons (i.e. not just terminal exons). Do we want to remove? Maybe not necessary, don't know if tools try to handle this uniformly. GT is just 3'seq (mostly) so should define the PASs. May inflate the number of FNs if a tool removes these kinds of sites.

In [45]:
def assign_id(gr, cols_to_cat=["Chromosome","Start","End","Strand"], sep_char=":", out_col="pas_id"):
    '''
    '''
    
    return gr.assign(out_col,
                     lambda df: df[cols_to_cat[0]].str.cat(df[cols_to_cat[1:]].astype(str),
                                                           sep=sep_char
                                                           )
                    )

m_te_mayr_cd_r3_tpm_filt = assign_id(m_te_mayr_cd_r3_tpm_filt,
                                     cols_to_cat=["Chromosome","Start","End","Strand", "gene_id"],
                                     out_col="te_id")
m_te_mayr_cd_r3_tpm_filt

+--------------+-----------+-----------+--------------+-------+
| Chromosome   | Start     | End       | Strand       | +2    |
| (category)   | (int32)   | (int32)   | (category)   | ...   |
|--------------+-----------+-----------+--------------+-------|
| chr1         | 169804074 | 169804386 | +            | ...   |
| chr1         | 169807790 | 169807837 | +            | ...   |
| chr1         | 169821678 | 169821719 | +            | ...   |
| chr1         | 169852789 | 169854080 | +            | ...   |
| ...          | ...       | ...       | ...          | ...   |
| chrY         | 11153857  | 11154070  | -            | ...   |
| chrY         | 11155653  | 11155884  | -            | ...   |
| chrY         | 11162449  | 11163137  | -            | ...   |
| chrY         | 12661308  | 12662124  | -            | ...   |
+--------------+-----------+-----------+--------------+-------+
Stranded PyRanges object has 48,902 rows and 6 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
2 hidden columns: gene_id, te_id

Have a set of TEs that pass expression filter, now can filter the GT PAS for those overlapping these TEs


# 3. Select TEs that have APA (>= 2 polyA sites)

Want to also track:
1. Number of PAS that are lost to filtering



In [47]:
gt_pas = pr.read_bed("data/gt/Mayr_CD5B_R3.SRR6795684.3seq.hg38.bed")
gt_pas

+--------------+-----------+-----------+-----------------------+-------+
| Chromosome   | Start     | End       | Name                  | +2    |
| (category)   | (int32)   | (int32)   | (object)              | ...   |
|--------------+-----------+-----------+-----------------------+-------|
| 1            | 629238    | 629239    | 1:629219-629259:+     | ...   |
| 1            | 630540    | 630541    | 1:630520-630563:+     | ...   |
| 1            | 634373    | 634374    | 1:634350-634378:+     | ...   |
| 1            | 1014536   | 1014537   | 1:1014512-1014544:+   | ...   |
| ...          | ...       | ...       | ...                   | ...   |
| Y            | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |
| Y            | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |
| Y            | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |
| Y            | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |
+--------------+-----------+-----------+-----------------------+-------+
Stranded PyRanges object has 13,343 rows and 6 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
2 hidden columns: Score, Strand

In [48]:
gt_pas.Chromosome = "chr" + gt_pas.Chromosome.astype(str)
gt_pas

+--------------+-----------+-----------+-----------------------+-------+
| Chromosome   | Start     | End       | Name                  | +2    |
| (category)   | (int32)   | (int32)   | (object)              | ...   |
|--------------+-----------+-----------+-----------------------+-------|
| chr1         | 629238    | 629239    | 1:629219-629259:+     | ...   |
| chr1         | 630540    | 630541    | 1:630520-630563:+     | ...   |
| chr1         | 634373    | 634374    | 1:634350-634378:+     | ...   |
| chr1         | 1014536   | 1014537   | 1:1014512-1014544:+   | ...   |
| ...          | ...       | ...       | ...                   | ...   |
| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |
| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |
| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |
| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |
+--------------+-----------+-----------+-----------------------+-------+
Stranded PyRanges object has 13,343 rows and 6 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
2 hidden columns: Score, Strand

In [53]:
gt_pas = assign_id(gt_pas)
gt_pas

+--------------+-----------+-----------+-----------------------+-------+
| Chromosome   | Start     | End       | Name                  | +3    |
| (category)   | (int32)   | (int32)   | (object)              | ...   |
|--------------+-----------+-----------+-----------------------+-------|
| chr1         | 629238    | 629239    | 1:629219-629259:+     | ...   |
| chr1         | 630540    | 630541    | 1:630520-630563:+     | ...   |
| chr1         | 634373    | 634374    | 1:634350-634378:+     | ...   |
| chr1         | 1014536   | 1014537   | 1:1014512-1014544:+   | ...   |
| ...          | ...       | ...       | ...                   | ...   |
| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |
| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |
| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |
| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |
+--------------+-----------+-----------+-----------------------+-------+
Stranded PyRanges object has 13,343 rows and 7 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
3 hidden columns: Score, Strand, pas_id

In [20]:
m_te_mayr_cd_r3_tpm_filt.Chromosome.unique().tolist()

['chr1',
 'chr2',
 'chr3',
 'chr4',
 'chr5',
 'chr6',
 'chr7',
 'chr8',
 'chr9',
 'chr10',
 'chr11',
 'chr12',
 'chr13',
 'chr14',
 'chr15',
 'chr16',
 'chr17',
 'chr18',
 'chr19',
 'chr20',
 'chr21',
 'chr22',
 'chrM',
 'chrX',
 'chrY']

In [54]:
n_gt_pas = gt_pas.pas_id.nunique()
n_gt_pas

13343

How many GT PAS do not overlap filtered terminal exons?

In [56]:

gt_pas = pr.PyRanges(gt_pas.as_df(), int64=True)
m_te_mayr_cd_r3_tpm_filt = pr.PyRanges(m_te_mayr_cd_r3_tpm_filt.as_df(), int64=True)

# pr.join often complains if the int dtypes of coordinates are different between PyRanges/not int64
n_gt_pas_no_olap = (gt_pas.join(m_te_mayr_cd_r3_tpm_filt,
             how="left", strandedness="same")
 .subset(lambda df: df["te_id"] == "-1")
 .pas_id.nunique()
)

n_gt_pas_no_olap

5051

In [57]:
n_gt_pas_no_olap / n_gt_pas

0.37855055085063327

~ 38 % do not overlap with filtered TEs...

How many PAS occur on TEs by themselves? i.e. only a single GT PAS for that TE

In [62]:
n_gt_pas_single = (m_te_mayr_cd_r3_tpm_filt.count_overlaps(gt_pas, strandedness="same")
 .subset(lambda df: df["NumberOverlaps"] == 1)
 # Since each TE only has one overlapping pas, can use te_id count as proxy for pas IDs
 .te_id.nunique()
)

n_gt_pas_single

4730

In [63]:
n_gt_pas_single / n_gt_pas

0.35449299258037925

In [71]:
n_gt_pas_min2 = n_gt_pas - n_gt_pas_no_olap - n_gt_pas_single
n_gt_pas_min2

3562

Only 3562 / 13343 PAS are overlapping filtered TEs for our analysis. This number of PAS will shrink as select 2 representative PAS for TEs that have > 2

In [66]:
# TEs vs number of overlapping PAS
# Again perhaps not so surprising to see large propn not having APA
## Isoform not necessarily expressed, TE only has 1 pas etc.
(m_te_mayr_cd_r3_tpm_filt.count_overlaps(gt_pas, strandedness="same")
 .as_df()["NumberOverlaps"].describe(percentiles = [i * 0.01 for i in range(0,100,5)]))

count    48902.00000
mean         0.17077
std          0.55754
min          0.00000
0%           0.00000
5%           0.00000
10%          0.00000
15%          0.00000
20%          0.00000
25%          0.00000
30%          0.00000
35%          0.00000
40%          0.00000
45%          0.00000
50%          0.00000
55%          0.00000
60%          0.00000
65%          0.00000
70%          0.00000
75%          0.00000
80%          0.00000
85%          0.00000
90%          1.00000
95%          1.00000
max         34.00000
Name: NumberOverlaps, dtype: float64

Now subset to valid TEs (at least two PAS)

In [67]:
# TEs have at least 2 overlapping PAS
valid_tes = (m_te_mayr_cd_r3_tpm_filt.count_overlaps(gt_pas, strandedness="same")
             .subset(lambda df: df["NumberOverlaps"] >= 2))

valid_tes

+--------------+-----------+-----------+--------------+-------+
| Chromosome   | Start     | End       | Strand       | +3    |
| (category)   | (int64)   | (int64)   | (category)   | ...   |
|--------------+-----------+-----------+--------------+-------|
| chr1         | 24468985  | 24472976  | +            | ...   |
| chr1         | 11840142  | 11843143  | +            | ...   |
| chr1         | 150158818 | 150160065 | +            | ...   |
| chr1         | 12508892  | 12512047  | +            | ...   |
| ...          | ...       | ...       | ...          | ...   |
| chrX         | 107713220 | 107714749 | -            | ...   |
| chrX         | 1386151   | 1386759   | -            | ...   |
| chrX         | 72272041  | 72272772  | -            | ...   |
| chrX         | 149477103 | 149483218 | -            | ...   |
+--------------+-----------+-----------+--------------+-------+
Stranded PyRanges object has 1,416 rows and 7 columns from 24 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
3 hidden columns: gene_id, te_id, NumberOverlaps

In [68]:
print(f"Number of TEs with at least 2 GT PAS - {valid_tes.te_id.nunique()}")

Number of TEs with at least 2 GT PAS - 1416


In [69]:
# Now subset to PAS overlapping these TEs
gt_pas_tes = gt_pas.join(valid_tes, strandedness="same")

gt_pas_tes

+--------------+-----------+-----------+-----------------------+-------+
| Chromosome   | Start     | End       | Name                  | +9    |
| (category)   | (int64)   | (int64)   | (object)              | ...   |
|--------------+-----------+-----------+-----------------------+-------|
| chr1         | 2310137   | 2310138   | 1:2310102-2310140:+   | ...   |
| chr1         | 2310180   | 2310181   | 1:2310180-2310181:+   | ...   |
| chr1         | 2310205   | 2310206   | 1:2310183-2310220:+   | ...   |
| chr1         | 2403215   | 2403216   | 1:2403190-2403237:+   | ...   |
| ...          | ...       | ...       | ...                   | ...   |
| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |
| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |
| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |
| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |
+--------------+-----------+-----------+-----------------------+-------+
Stranded PyRanges object has 3,621 rows and 13 columns from 24 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
9 hidden columns: Score, Strand, pas_id, Start_b, End_b, Strand_b, gene_id, ... (+ 2 more.)


## 4. Select two representative PAS for TEs with > 2 PASs

If TE has 2 PAS - those are the selected PAS for that TE.
All other TEs - need to select two representative PAS, Joseph's suggestion was to take the two most highly expressed.



In [70]:
# Apparently the rank approach is much faster if we're worried about that...
# https://stackoverflow.com/questions/62157558/filter-for-rows-with-n-largest-values-for-each-group
# Score is the TPM expression of a PAS
# Select two highest expressed PASs for genes 

gt_pas_tes_2rep = (gt_pas_tes.subset(lambda df: df["NumberOverlaps"] > 2)
 .apply(lambda df: df.groupby("te_id").apply(lambda x: x.nlargest(2, "Score")).reset_index(drop=True))
)

gt_pas_tes_2rep

+--------------+-----------+-----------+-------------------------+-------+
| Chromosome   | Start     | End       | Name                    | +9    |
| (category)   | (int64)   | (int64)   | (object)                | ...   |
|--------------+-----------+-----------+-------------------------+-------|
| chr1         | 101241516 | 101241517 | 1:101241488-101241534:+ | ...   |
| chr1         | 101241381 | 101241382 | 1:101241358-101241382:+ | ...   |
| chr1         | 108700364 | 108700365 | 1:108700340-108700378:+ | ...   |
| chr1         | 108701192 | 108701193 | 1:108701170-108701193:+ | ...   |
| ...          | ...       | ...       | ...                     | ...   |
| chrX         | 107713326 | 107713327 | X:107713301-107713352:- | ...   |
| chrX         | 107713236 | 107713237 | X:107713236-107713250:- | ...   |
| chrX         | 1386250   | 1386251   | X:1386225-1386271:-     | ...   |
| chrX         | 1386151   | 1386152   | X:1386124-1386172:-     | ...   |
+--------------+-----------+-----------+-------------------------+-------+
Stranded PyRanges object has 886 rows and 13 columns from 23 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
9 hidden columns: Score, Strand, pas_id, Start_b, End_b, Strand_b, gene_id, te_id, ... (+ 1 more.)

In [72]:
# Combine selected 2 with other PAS on TEs with only 2 PAS to get final list of GT PAS
gt_pas_tes_rep = pr.concat([gt_pas_tes.subset(lambda df: df["NumberOverlaps"] == 2), gt_pas_tes_2rep])
gt_pas_tes_rep

+--------------+-----------+-----------+-----------------------+-------+
| Chromosome   | Start     | End       | Name                  | +9    |
| (category)   | (int64)   | (int64)   | (object)              | ...   |
|--------------+-----------+-----------+-----------------------+-------|
| chr1         | 11843045  | 11843046  | 1:11843016-11843046:+ | ...   |
| chr1         | 11843127  | 11843128  | 1:11843102-11843143:+ | ...   |
| chr1         | 12013470  | 12013471  | 1:12013468-12013476:+ | ...   |
| chr1         | 12013501  | 12013502  | 1:12013488-12013523:+ | ...   |
| ...          | ...       | ...       | ...                   | ...   |
| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |
| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |
| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |
| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |
+--------------+-----------+-----------+-----------------------+-------+
Stranded PyRanges object has 2,832 rows and 13 columns from 24 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
9 hidden columns: Score, Strand, pas_id, Start_b, End_b, Strand_b, gene_id, ... (+ 2 more.)

In [74]:
n_gt_pas_rep = gt_pas_tes_rep.pas_id.nunique()
n_gt_pas_rep

2818

Should be 2 PAS for every TE, but 1416 * 2 != 2818...

In [75]:
# Label as proximal / distal - assign region number 1 (proximal) & 2 (distal)
gt_pas_tes_rep = add_region_number(gt_pas_tes_rep.assign("Feature",
                                                         lambda x: pd.Series(["pas"]*len(x.index))),
                                   id_col="te_id",
                                   out_col="pas_number",
                                   feature_key="pas").drop("Feature")

gt_pas_tes_rep

+--------------+-----------+-----------+-----------------------+-------+
| Chromosome   | Start     | End       | Name                  | +10   |
| (category)   | (int64)   | (int64)   | (object)              | ...   |
|--------------+-----------+-----------+-----------------------+-------|
| chr1         | 2310137   | 2310138   | 1:2310102-2310140:+   | ...   |
| chr1         | 2310205   | 2310206   | 1:2310183-2310220:+   | ...   |
| chr1         | 2403215   | 2403216   | 1:2403190-2403237:+   | ...   |
| chr1         | 2405434   | 2405435   | 1:2405417-2405448:+   | ...   |
| ...          | ...       | ...       | ...                   | ...   |
| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |
| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |
| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |
| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |
+--------------+-----------+-----------+-----------------------+-------+
Stranded PyRanges object has 2,832 rows and 14 columns from 24 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
10 hidden columns: Score, Strand, pas_id, Start_b, End_b, Strand_b, gene_id, ... (+ 3 more.)

In [76]:
# Calculate fractional relative expression for each PAS (vs all PAS on TE)
gt_pas_tes_rep = gt_pas_tes_rep.assign("rel_exp",
                                       lambda df: df.groupby("te_id", group_keys=False)["Score"].apply(lambda x: x / x.sum())
                                       )

gt_pas_tes_rep

+--------------+-----------+-----------+-----------------------+-------+
| Chromosome   | Start     | End       | Name                  | +11   |
| (category)   | (int64)   | (int64)   | (object)              | ...   |
|--------------+-----------+-----------+-----------------------+-------|
| chr1         | 2310137   | 2310138   | 1:2310102-2310140:+   | ...   |
| chr1         | 2310205   | 2310206   | 1:2310183-2310220:+   | ...   |
| chr1         | 2403215   | 2403216   | 1:2403190-2403237:+   | ...   |
| chr1         | 2405434   | 2405435   | 1:2405417-2405448:+   | ...   |
| ...          | ...       | ...       | ...                   | ...   |
| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |
| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |
| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |
| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |
+--------------+-----------+-----------+-----------------------+-------+
Stranded PyRanges object has 2,832 rows and 15 columns from 24 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
11 hidden columns: Score, Strand, pas_id, Start_b, End_b, Strand_b, gene_id, ... (+ 4 more.)

# Match predictions to filtered ground truth

- Need to assign pred PAS to the correct TE - info not present in prediction BED files

**Really need to tackle problem of multiple predictions matching GT  here (due to tools running tx-by-tx)**

My initial suggestion was to take the two sites that have the 'best match' to ground truth. i.e. Take all PAS assigned to that composite/merged TE, and select the two isoforms that have the best match to the GT, in order of (both match, one of proximal/distal matches, no matches) With way I've named DaPars2 Name field there isn't an out of the box way to know which predictions have come from the same TE frame (would just have to extract the transcript ID). Since DaPars tools calc rel expression tx by tx, there could be a case where the proximal and distal sites match between different transcript models. This could mean that the predicted relative expressions do not sum to 1.


Could probably get around this by deciding on one of prox/distal relative expr to evaluate/compare against. In this case if have multiple predicted pas in a given window, could either take the site with the smallest error/difference to GT relative expression, or take sth like the mean/median of the predicted matched sites rel exp. If did it this way, would maybe be good to try both ways around (i.e. calc metrics relative to distal rel usage and then proximal rel usage).
         

In [77]:
pred_mayr_cd5_dapars2 = pr.read_bed("data/pred/Mayr_CD5B_R3_DaPars2_04.bed")
pred_mayr_cd5_dapars = pr.read_bed("data/pred/Mayr_CD5B_R3_relative_usage_quantification_output.bed")


# Name is not necessarily unique,to be safe (if need it) assign a ID for each PAS
pred_mayr_cd5_dapars2 = assign_id(pred_mayr_cd5_dapars2)
pred_mayr_cd5_dapars = assign_id(pred_mayr_cd5_dapars)

pred_mayr_cd5_dapars2

+--------------+-----------+-----------+-------+
| Chromosome   | Start     | End       | +4    |
| (category)   | (int32)   | (int32)   | ...   |
|--------------+-----------+-----------+-------|
| chr1         | 28453784  | 28453785  | ...   |
| chr1         | 51788223  | 51788224  | ...   |
| chr1         | 212103089 | 212103090 | ...   |
| chr1         | 202439280 | 202439281 | ...   |
| ...          | ...       | ...       | ...   |
| chrY         | 2881682   | 2881683   | ...   |
| chrY         | 19707181  | 19707182  | ...   |
| chrY         | 13234580  | 13234581  | ...   |
| chrY         | 19705424  | 19705425  | ...   |
+--------------+-----------+-----------+-------+
Stranded PyRanges object has 109,654 rows and 7 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
4 hidden columns: Name, Score, Strand, pas_id

In [78]:
pred_mayr_cd5_dapars

+--------------+-----------+-----------+-------+
| Chromosome   | Start     | End       | +4    |
| (category)   | (int32)   | (int32)   | ...   |
|--------------+-----------+-----------+-------|
| chr1         | 151034007 | 151034008 | ...   |
| chr1         | 6092096   | 6092097   | ...   |
| chr1         | 156800811 | 156800812 | ...   |
| chr1         | 111766684 | 111766685 | ...   |
| ...          | ...       | ...       | ...   |
| chrY         | 19705682  | 19705683  | ...   |
| chrY         | 2883109   | 2883110   | ...   |
| chrY         | 13323167  | 13323168  | ...   |
| chrY         | 2881682   | 2881683   | ...   |
+--------------+-----------+-----------+-------+
Stranded PyRanges object has 74,184 rows and 7 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
4 hidden columns: Name, Score, Strand, pas_id

As a dirty readout of potential amount of duplication of PAS due to multiple TE isoforms, see how many times identical pas_ids occur

Of course pas could differ very slightly if specific coords of TE isoforms are different & essentially be the same PAS

In [79]:
print("Fraction of predicted pas_id that are duplicated (i.e. identical coordinates)")
print("DaPars")
print(len(pred_mayr_cd5_dapars.pas_id.value_counts().loc[lambda x: x > 1]) / pred_mayr_cd5_dapars.pas_id.nunique())
print("DaPars2")
print(len(pred_mayr_cd5_dapars2.pas_id.value_counts().loc[lambda x: x > 1]) / pred_mayr_cd5_dapars2.pas_id.nunique())

Fraction of predicted pas_id that are duplicated (i.e. identical coordinates)
DaPars
0.09987056774420351
DaPars2
0.1019118652900368


~ 10 % of polyA sites are identical

In [84]:
# From a TE perspective - how many have >2 overlapping prediction sites?
# In the case of DaPars, this could mean overlapping terminal exon isoforms between different transcript IDs
##  DaPars & DaPars2 use a 2-site model, so number of predicted sites can never exceed expected 2
# how many have no overlapping predictions (or )

# Need to convert back to int32 otherwise get following error...
# ValueError: Buffer dtype mismatch, expected 'const int32_t' but got 'long

temp_dapars = pr.PyRanges(valid_tes.as_df(), int64=False).count_overlaps(pr.PyRanges(pred_mayr_cd5_dapars.as_df(), int64=False),
                         strandedness="same",
                         keep_nonoverlapping=True,
                         overlap_col="NumberOverlaps_pred")
temp_dapars2 = pr.PyRanges(valid_tes.as_df(), int64=False).count_overlaps(pr.PyRanges(pred_mayr_cd5_dapars2.as_df(), int64=False),
                         strandedness="same",
                         keep_nonoverlapping=True,
                         overlap_col="NumberOverlaps_pred")


print(temp_dapars.subset(lambda df: df["NumberOverlaps_pred"] < 2).NumberOverlaps_pred.value_counts())
temp_dapars2.subset(lambda df: df["NumberOverlaps_pred"] < 2).NumberOverlaps_pred.value_counts()


0    247
Name: NumberOverlaps_pred, dtype: int64


0    149
Name: NumberOverlaps_pred, dtype: int64

In [85]:
print("Fraction of curated GT terminal exons that have no predicted APA (i.e. < 2 PAS)")
print("DaPars")
print(len(temp_dapars.subset(lambda df: df["NumberOverlaps_pred"] < 2).NumberOverlaps_pred) / temp_dapars.te_id.nunique())
print("DaPars2")
print(len(temp_dapars2.subset(lambda df: df["NumberOverlaps_pred"] < 2).NumberOverlaps_pred) / temp_dapars2.te_id.nunique())

Fraction of curated GT terminal exons that have no predicted APA (i.e. < 2 PAS)
DaPars
0.17443502824858756
DaPars2
0.10522598870056497


Above is Joseph's suggestion for 'FNs'

In [86]:
# Assign predicted PAS to GT TEs
keep_cols = pred_mayr_cd5_dapars.columns.tolist() + ["te_id"]
pred_mayr_cd5_dapars = (pr.PyRanges(pred_mayr_cd5_dapars.as_df(),int64=True).join(pr.PyRanges(valid_tes.as_df(),int64=True),
                                                          how="left",
                                                          strandedness="same", 
                                                          suffix="_te"))[keep_cols]

pred_mayr_cd5_dapars2 = (pr.PyRanges(pred_mayr_cd5_dapars2.as_df(),int64=True).join(pr.PyRanges(valid_tes.as_df(),
                                                                                                int64=True),
                                                          how="left",
                                                          strandedness="same", 
                                                          suffix="_te"))[keep_cols]

pred_mayr_cd5_dapars2

+--------------+-----------+-----------+-------+
| Chromosome   | Start     | End       | +5    |
| (category)   | (int64)   | (int64)   | ...   |
|--------------+-----------+-----------+-------|
| chr1         | 28496684  | 28496685  | ...   |
| chr1         | 53048234  | 53048235  | ...   |
| chr1         | 161120697 | 161120698 | ...   |
| chr1         | 44778726  | 44778727  | ...   |
| ...          | ...       | ...       | ...   |
| chrY         | 13234579  | 13234580  | ...   |
| chrY         | 2881682   | 2881683   | ...   |
| chrY         | 19707181  | 19707182  | ...   |
| chrY         | 13234580  | 13234581  | ...   |
+--------------+-----------+-----------+-------+
Stranded PyRanges object has 109,746 rows and 8 columns from 25 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.
5 hidden columns: Name, Score, Strand, pas_id, te_id

In [87]:
print("Fraction of predicted PAS that do not overlap with curated terminal exons")
print("DaPars")
print(pred_mayr_cd5_dapars.subset(lambda df: df["te_id"] == "-1").pas_id.nunique() / pred_mayr_cd5_dapars.pas_id.nunique())
print("DaPars2")
print(pred_mayr_cd5_dapars2.subset(lambda df: df["te_id"] == "-1").pas_id.nunique() / pred_mayr_cd5_dapars2.pas_id.nunique())

Fraction of predicted PAS that do not overlap with curated terminal exons
DaPars
0.8895191831386523
DaPars2
0.911604729179559


Taking this approach of curating GT data means excluding vast majority of predicted PAS for both DaPars tools. If we considered predictions within a window to be the same PAS, that may reduce the number
